### PR TITLE
feat(Encryption): inline & customize URLcrypt

### PR DESCRIPTION
**Pull request by danielma**
URLcrypt is an important part of the functionality of `pco-url`, but
it has 2 major drawbacks.

1. It uses a module instance variable for a shared key across. This
   means it can't be used with more than one key
2. Based on issue 1, it's not threadsafe with multiple keys (which is
   a thing we currently need to support in accounts)

This PR takes a few steps, but the high level is to inline `URLcrypt`
into `PCO::URL::Encryption` and then to make it threadsafe by not
allowing apps to change the key, and allowing multiple keys by passing
an explicit key to an encrypt/decrypt cycle

### CODE REVIEW

**Comment by wassimk** [comment]
This looks really great! Thanks for the clean effort 💯

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+version: 2 # use CircleCI 2.0
```
Comment (molawson) [inline .circleci/config.yml]: 👏 💯 🙇

File: spec/pco/url/encryption_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
+require "spec_helper"
+
+describe PCO::URL::Encryption do
```
Comment (molawson) [inline spec/pco/url/encryption_spec.rb]: I don't want to slow this down unnecessarily, but do you think it would be worth adding test cases from the [gem](https://github.com/cheerful/URLcrypt/blob/v0.1.1/test/URLcrypt_test.rb) to cover the more edge-casey parts of this encryption?  That's the sort of stuff that I'm way out of my league on.  But it also seems like it's important 🤷‍♂️

File: lib/pco/url/encryption.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+          [
+            (0..(n - 1)).to_a.reverse.collect { |i| TABLE[(c >> i * 5) & 0x1f].chr },
+            ("=" * (8 - n))
+          ]
+        end
+      end
+
+      class << self
+        def default_key=(key)
+          raise(<<~ERR) if defined?(@default_key) && @default_key != key
```
Comment (molawson) [inline lib/pco/url/encryption.rb]: I had no idea this syntax was even valid Ruby with the heredoc below it and all 🤯

File: lib/pco/url/encryption.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            Don't change the default_key during the life of a process.
+            This configuration variable is meant to be set once on boot and not
+            touched again.
+
+            If your app needs to use multiple keys, pass them in explicitly to `encrypt`
+            and `decrypt`
+          ERR
+
+          @default_key = key
```
Comment (molawson) [inline lib/pco/url/encryption.rb]: I love the expressive error message above!  I think the conditional covers us in the cases where the key is set on startup. But the module instance variable still scares me a little.  What do you think about using something like `Thread.current` (or `RequestStore`) as the storage mechanism here?  The idea being that it's not much effort, but we get to push off the real thread-safety to a tool that's made for it instead of having to think of the edge cases ourselves.

File: spec/pco/url/encryption_spec.rb (hunk 2 above)
Comment (danielma) [reply spec/pco/url/encryption_spec.rb]: yeah that totally makes sense. I had kinda spaced out on it because I just assume this code works and no-one will touch it 😄

File: lib/pco/url/encryption.rb (hunk 3 above)
Comment (danielma) [reply lib/pco/url/encryption.rb]: yeah uh... it's really weird. the kind of stuff you learn when you spend a free week working on a ruby formatter

File: lib/pco/url/encryption.rb (hunk 4 above)
Comment (danielma) [reply lib/pco/url/encryption.rb]: Honestly, I'm not sure what that would get us. This is specifically designed to ensure that the default key doesn't change, even across threads. So storing in `thread.current` kind of breaks the model, in that one thread or another could have different value for `default_key`, and I don't want that. 

In my mind, the "thread safety" of his PR comes from the fact that the key can't change. `Thread.current`/`RequestStore` (in my mind) are more about "thread isolation".

If I'm missing something or off my rocker, please let me know

Comment (molawson) [reply lib/pco/url/encryption.rb]: No, I see that.  I'm good keeping it as-is.

File: spec/pco_url_spec.rb [hunk 5]
```diff
@@ -14,8 +14,12 @@
 
 describe PCO::URL do
   before :all do
-    URLcrypt.key = "984e9002e680dc9b9c2556434c47f7e4782191f52063277901e4a009797652e0" \
-                   "8f28be069dfb4d4a1e3c9ab09fedab59be2c9b6486748bf44030182815ee4987"
+    PCO::URL::Encryption.default_key =
+      "\xF7\xFE\x99I\x1EkO\xD4\xD3\v\x96\x8A\b\x17\xD8m\x01jr\x8F\xA0L.\xB3\xF3\x12\xD7c\x16\xA8\xD0."
```
Comment (bensie) [inline spec/pco_url_spec.rb]: Unless we intend to rotate both current keys, we should probably continue testing keys formatted the way ours is in consul. I remember you mentioning that this is the expected input for OpenSSL, but maybe what we're using is still technically ok? Or partially truncated? 🤷‍♂️

**Comment by danielma** [comment]
FWIW These values come direct from `SecureRandom.random_bytes(32)`, which
is what Rails is using to generate these kind of keys. That’s the reason
they have all the weird bytes. I have no particular feeling about these vs
all ascii characters
